### PR TITLE
[EXP] add sourmash distance estimation

### LINKED ISSUE #1242: implement pairwise evolutionary distance/ANI output from sourmash
what this means in practical concerns is not clear, but...

cc @bluegenes

### PR DESCRIPTION
**Pull request by bluegenes**
Implement a minimal version of https://github.com/KoslickiLab/mutation-rate-ci-calculator to estimate distance from scaled containment and jaccard.

See https://www.biorxiv.org/content/10.1101/2022.01.11.475870v1

May fix https://github.com/sourmash-bio/sourmash/issues/1242

- actually use distance utilities within sourmash:
  - [x] add signature methods for jaccard, containment, max_containment --> ANI
  - [x] `compare`
  - [x] `prefetch`
  - [x] `gather` 
  - [x] `search`
 
add'l to do:
- [x] Add functions for edge case detection/reporting (no sketch overlap, 100% sketch overlap by chance).
- [ ] decide when/where to report confidence intervals.
  - Report in `search`, `prefetch`, `gather`, NOT in `compare`. This maintains compatibility with `sourmash plot`.
- [x] handle sketch downsampling in `signature` ANI functions
- [x] add `confidence` param to `minhash`, `signature` ANI functions to allow tuning confidence level. Defaults to `0.95`.
- [x] add `--ani-threshold` to `tax genome`
- [ ] ensure @dkoslicki and @mahmudhera are in sourmash authors list (for JOSS papers)
- [ ] understand benchmarking issues

**Questions/Thoughts:**
- Do we also want to report additional metrics, e.g. fraction genome matched?
- Is there any issue with directly reporting the estimated ANI's (1-distance)?  Should we enable both "distance" and ANI and let user choose?

**NOTE:** ANI estimation is not backwards compatible with previous prefetch/gather results (needs additional columns).

### CODE REVIEW

**Comment by ctb** [comment]
lookin' good. do you want comments now or do you want to wait until it's more fully baked?

**Comment by bluegenes** [comment]
@ctb Comments now would be good! I'm done adding things for now, just gonna go take it for a spin :)

File: src/sourmash/commands.py [hunk 1]
```diff
@@ -138,12 +147,12 @@ def compare(args):
 
     labeltext = [str(item) for item in siglist]
     if args.containment:
-        similarity = compare_serial_containment(siglist)
+        similarity = compare_serial_containment(siglist, return_ANI=return_ani)
```
Comment (ctb) [inline src/sourmash/commands.py]:
```suggestion
        similarity = compare_serial_containment(siglist, return_ani=return_ani)
```

File: src/sourmash/commands.py [hunk 2]
```diff
@@ -138,12 +147,12 @@ def compare(args):
...
     elif args.max_containment:
-        similarity = compare_serial_max_containment(siglist)
+        similarity = compare_serial_max_containment(siglist, return_ANI=return_ani)
```
Comment (ctb) [inline src/sourmash/commands.py]:
```suggestion
        similarity = compare_serial_max_containment(siglist, return_ani=return_ani)
```

File: src/sourmash/commands.py [hunk 3]
```diff
@@ -138,12 +147,12 @@ def compare(args):
...
     else:
         similarity = compare_all_pairs(siglist, args.ignore_abundance,
-                                       n_jobs=args.processes)
+                                       n_jobs=args.processes, return_ANI=return_ani)
```
Comment (ctb) [inline src/sourmash/commands.py]:
```suggestion
                                       n_jobs=args.processes, return_ani=return_ani)
```

File: src/sourmash/commands.py [hunk 4]
```diff
@@ -792,12 +804,14 @@ def gather(args):
 
     # save CSV?
     if found and args.output:
-        fieldnames = ['intersect_bp', 'f_orig_query', 'f_match',
-                      'f_unique_to_query', 'f_unique_weighted',
-                      'average_abund', 'median_abund', 'std_abund', 'name',
-                      'filename', 'md5', 'f_match_orig', 'unique_intersect_bp',
-                      'gather_result_rank', 'remaining_bp',
-                      'query_filename', 'query_name', 'query_md5', 'query_bp']
+        fieldnames = GatherResult._fields
```
Comment (ctb) [inline src/sourmash/commands.py:815]: 👍 !!!

File: src/sourmash/compare.py [hunk 5]
```diff
@@ -9,7 +9,7 @@
 from sourmash.np_utils import to_memmap
 
 
-def compare_serial(siglist, ignore_abundance, downsample=False):
+def compare_serial(siglist, ignore_abundance, downsample=False, return_ANI=False):
```
Comment (ctb) [inline src/sourmash/compare.py]:
```suggestion
def compare_serial(siglist, ignore_abundance, downsample=False, return_ani=False):
```

File: src/sourmash/compare.py [hunk 6]
```diff
@@ -77,22 +84,30 @@ def compare_serial_max_containment(siglist, downsample=False):
...
+                                                         downsample=downsample)[0]
+            else:
+                containments[i][j] = siglist[j].max_containment(siglist[i],
                                                          downsample=downsample)
 
     return containments
 
 
-def similarity_args_unpack(args, ignore_abundance, downsample):
+def similarity_args_unpack(args, ignore_abundance, downsample, return_ANI=False):
```
Comment (ctb) [inline src/sourmash/compare.py]:
```suggestion
def similarity_args_unpack(args, ignore_abundance, downsample, return_ani=False):
```

File: src/sourmash/compare.py [hunk 7]
```diff
@@ -77,22 +84,30 @@ def compare_serial_max_containment(siglist, downsample=False):
...
     """Helper function to unpack the arguments. Written to use in pool.imap
     as it can only be given one argument."""
     sig1, sig2 = args
-    return sig1.similarity(sig2,
+    if return_ANI:
```
Comment (ctb) [inline src/sourmash/compare.py]:
```suggestion
    if return_ani:
```

File: src/sourmash/compare.py [hunk 8]
```diff
@@ -77,22 +84,30 @@ def compare_serial_max_containment(siglist, downsample=False):
...
+        return sig1.jaccard_ani(sig2,
+                           downsample=downsample)[0]
+    else:
+        return sig1.similarity(sig2,
                            ignore_abundance=ignore_abundance,
                            downsample=downsample)
 
 
-def get_similarities_at_index(index, ignore_abundance, downsample, siglist):
+def get_similarities_at_index(index, ignore_abundance, downsample, siglist, return_ANI=False):
```
Comment (ctb) [inline src/sourmash/compare.py]:
```suggestion
def get_similarities_at_index(index, ignore_abundance, downsample, siglist, return_ani=False):
```

File: src/sourmash/compare.py [hunk 9]
```diff
@@ -114,14 +129,14 @@ def get_similarities_at_index(index, ignore_abundance, downsample, siglist):
     sig_iterator = itertools.product([siglist[index]], siglist[index + 1:])
     func = partial(similarity_args_unpack,
                    ignore_abundance=ignore_abundance,
-                   downsample=downsample)
+                   downsample=downsample, return_ANI=return_ANI)
```
Comment (ctb) [inline src/sourmash/compare.py]:
```suggestion
                   downsample=downsample, return_ani=return_ani)
```

File: src/sourmash/compare.py [hunk 10]
```diff
@@ -114,14 +129,14 @@ def get_similarities_at_index(index, ignore_abundance, downsample, siglist):
...
     similarity_list = list(map(func, sig_iterator))
     notify(
         f"comparison for index {index} done in {time.time() - startt:.5f} seconds", end='\r')
     return similarity_list
 
 
-def compare_parallel(siglist, ignore_abundance, downsample, n_jobs):
+def compare_parallel(siglist, ignore_abundance, downsample, n_jobs, return_ANI=False):
```
Comment (ctb) [inline src/sourmash/compare.py]:
```suggestion
def compare_parallel(siglist, ignore_abundance, downsample, n_jobs, return_ani=False):
```

File: src/sourmash/compare.py [hunk 11]
```diff
@@ -163,7 +178,8 @@ def compare_parallel(siglist, ignore_abundance, downsample, n_jobs):
         get_similarities_at_index,
         siglist=siglist,
         ignore_abundance=ignore_abundance,
-        downsample=downsample)
+        downsample=downsample,
+        return_ANI=return_ANI)
```
Comment (ctb) [inline src/sourmash/compare.py]:
```suggestion
        return_ani=return_ani)
```

File: src/sourmash/compare.py [hunk 12]
```diff
@@ -198,7 +214,7 @@ def compare_parallel(siglist, ignore_abundance, downsample, n_jobs):
     return np.memmap(filename, dtype=np.float64, shape=(length_siglist, length_siglist))
 
 
-def compare_all_pairs(siglist, ignore_abundance, downsample=False, n_jobs=None):
+def compare_all_pairs(siglist, ignore_abundance, downsample=False, n_jobs=None, return_ANI=False):
```
Comment (ctb) [inline src/sourmash/compare.py]:
```suggestion
def compare_all_pairs(siglist, ignore_abundance, downsample=False, n_jobs=None, return_ani=False):
```

File: src/sourmash/commands.py (hunk 4 above)
Comment (bluegenes) [reply src/sourmash/commands.py:815]: oh yay! I wasn't sure if you wanted these explicitly for some reason 😂

File: src/sourmash/compare.py [hunk 13]
```diff
@@ -216,7 +232,7 @@ def compare_all_pairs(siglist, ignore_abundance, downsample=False, n_jobs=None):
     :return: np.array similarity matrix
     """
     if n_jobs is None or n_jobs == 1:
-        similarities = compare_serial(siglist, ignore_abundance, downsample)
+        similarities = compare_serial(siglist, ignore_abundance, downsample, return_ANI)
```
Comment (ctb) [inline src/sourmash/compare.py]:
```suggestion
        similarities = compare_serial(siglist, ignore_abundance, downsample, return_ani)
```

File: src/sourmash/compare.py [hunk 14]
```diff
@@ -216,7 +232,7 @@ def compare_all_pairs(siglist, ignore_abundance, downsample=False, n_jobs=None):
...
     else:
-        similarities = compare_parallel(siglist, ignore_abundance, downsample, n_jobs)
+        similarities = compare_parallel(siglist, ignore_abundance, downsample, n_jobs, return_ANI)
```
Comment (ctb) [inline src/sourmash/compare.py]:
```suggestion
        similarities = compare_parallel(siglist, ignore_abundance, downsample, n_jobs, return_ani)
```

**Comment by ctb** [comment]
> **Questions/Thoughts:**
> 
>     * Currently not reporting confidence interval in `compare`, as the results are stored as a single np matrix. How can we store + report these? Larger CSV output, rather than symmetric matrix? This would elim compatibility with current`plot`. What other issues would larger CSV bring about?

I _think_ leave it out for now. Would it be included in `search` output?
 
>     * Do we also want to report additional metrics, e.g. fraction genome matched? If yes, same issues as ^.

Same answer :). But we might want to NA or zero out the numbers where they are particularly sketch-y (heh)

>     * Is there any issue with directly reporting the estimated ANI's (1-distance)?  Should we enable both "distance" and ANI and let user choose?

ANI good, IMO!

File: src/sourmash/signature.py [hunk 15]
```diff
@@ -142,14 +143,48 @@ def jaccard(self, other):
         return self.minhash.similarity(other.minhash, ignore_abundance=True,
                                        downsample=False)
```
Comment (bluegenes) [inline src/sourmash/signature.py:145]: @ctb why does `jaccard` not allow downsampling?

**Comment by ctb** [comment]
On Mon, Jan 17, 2022 at 02:04:49PM -0800, Tessa Pierce Ward wrote:
> @bluegenes commented on this pull request.
> 
> > @@ -142,14 +143,48 @@ def jaccard(self, other):
>          return self.minhash.similarity(other.minhash, ignore_abundance=True,
>                                         downsample=False)
>  
> 
> @ctb why does `jaccard` not allow downsampling?

I can't think of a good reason off hand :). Probably just didn't implement.

File: src/sourmash/signature.py (hunk 15 above)
Comment (ctb) [reply src/sourmash/signature.py:145]: confirmed, no real reason as far as I can tell. if you felt like adding in a comment saying "there's no theoretical reason for this" it would be appropriate.

File: src/sourmash/index.py [hunk 16]
```diff
@@ -201,7 +201,7 @@ def search_abund(self, query, *, threshold=None, **kwargs):
                 raise TypeError("'search_abund' requires subject signatures with abundance information")
             score = query.similarity(subj)
             if score >= threshold:
-                matches.append(IndexSearchResult(score, subj, self.location))
```
Comment (bluegenes) [inline src/sourmash/index.py]: pretty sure this was a bug, @ctb, but pointing out so you can double check. New test (`test_search_ani_empty_abund`) in `test_sourmash.py` revealed it -- the `filename` entry in the results csv was empty (+ is correct with this change).

Comment (ctb) [reply src/sourmash/index.py]: agreed, and nice catch! I might try to swipe this fix and combine it with a fix for https://github.com/sourmash-bio/sourmash/issues/1804 if that's ok by you?

Comment (bluegenes) [reply src/sourmash/index.py]: Sure!

Comment (ctb) [reply src/sourmash/index.py]: redundantly fixed in https://github.com/sourmash-bio/sourmash/pull/1820; resolving this conversation even though that hasn't been merged yet :)

**Comment by bluegenes** [comment]
Occasionally running into the following warnings. To do: track down and understand why/when this is happening.

```
 RuntimeWarning: invalid value encountered in sqrt
  f2 = lambda pest: 2.0/(2- (1-pest)**ksize ) - 1 - z_alpha * sqrt(var_direct(pest)) - jaccard
/home/ntpierce/miniconda3/7cc50dccfe87e0e0e5a644dafd4899b7/lib/python3.9/site-packages/sourmash/distance_utils.py:145: RuntimeWarning: invalid value encountered in sqrt
  f1 = lambda pest: 2.0/(2- (1-pest)**ksize ) - 1 + z_alpha * sqrt(var_direct(pest)) - jaccard
```

Have you encountered this, @mahmudhera?

**Comment by mahmudhera** [comment]
Strangely, I did not. Then again, I must not have tested as extensively as you are doing. 

In any case, the theoretical analyses for Jaccard obtained from two FracMinHash sketches are not finished yet (and are not in the rxiv manuscript). I think I added it in the code to take it out for a spin. I must have made a very casual note to not trust this code, leading to this confusion. The code uses the ratio of expectations to approximate an expectation of ratios, and the rate of error has not yet been analyzed completely. 

In summary, using jaccard_to_distance is not theoretically analyzed yet, so I would recommend not using this. However, if you decide to only use the point estimate (and not the interval), that can be done quickly. If the analysis shows that the rate of error is not too large, we could obtain a point estimate of the mutation rate from Jaccard. Let me know @bluegenes if I should do this.

**Comment by bluegenes** [comment]
Hi @mahmudhera,

Thanks for the explanation! I hadn't realized where jaccard was relative to containment. On the plus side, I've been running both jaccard and containment for each of my comparisons, and the point estimates have been looking pretty similar.

> However, if you decide to only use the point estimate (and not the interval), that can be done quickly. If the analysis shows that the rate of error is not too large, we could obtain a point estimate of the mutation rate from Jaccard. Let me know @bluegenes if I should do this.

I would greatly appreciate this! Just the point estimate is great, intervals not needed.

**Comment by mahmudhera** [comment]
Excellent! The point estimate formula would essentially stay the same, I just want to do some theoretical analysis first, so that we have justification for using the formula. I will do some analyses in the next couple of days and give you an update on this.

**Comment by mahmudhera** [comment]
It looks like the point estimate that we have right now: it would only be reasonable to use when the ANI is large. Otherwise, the formula may work well by chance, but there is no theoretical guarantee. The exact range of ANI depends on other parameters. 

We can approach this in one of two ways: we can either keep the code as is, and besides the point estimate, we can return an estimated error. If the error is above some threshold, we can report that Jaccard->mutation_rate code cannot be trusted. Or, we can try to improve the error rate by solving for the point estimate differently, which may take a bit more work. I will try both locally, let me know which one of these you would prefer.

Meanwhile, do you think you can send the exact settings where the code failed? Sending the values of L, k, p, s for the cases where you got this RuntimeWarning would suffice.

**Comment by bluegenes** [comment]
thanks @mahmudhera!

> It looks like the point estimate that we have right now: it would only be reasonable to use when the ANI is large. Otherwise, the formula may work well by chance, but there is no theoretical guarantee. The exact range of ANI depends on other parameters.

ok, good to know!

> We can approach this in one of two ways: we can either keep the code as is, and besides the point estimate, we can return an estimated error. If the error is above some threshold, we can report that Jaccard->mutation_rate code cannot be trusted. Or, we can try to improve the error rate by solving for the point estimate differently, which may take a bit more work. I will try both locally, let me know which one of these you would prefer.

I think if it's possible to generate the jaccard ANI point estimate differently without too much extra work, that would be great. If not, the other option is ok -- I would probably just zero out the ANI estimate when the error is too high (using some default, user-modifiable threshold). If the error is much higher, we can recommend only containment --> ANI and provide jaccard equations with warnings. Alternatively, we could recommend using containment only.

>Meanwhile, do you think you can send the exact settings where the code failed? Sending the values of L, k, p, s for the cases where you got this RuntimeWarning would suffice.


I need to do some digging to get the exact parameters for when the warnings appeared -- will send over when I have them!

**Comment by bluegenes** [comment]
Relevant portion of a traceback for the RuntimeWarnings encountered w/ jaccard --> ANI

```
  File "/home/ntpierce/miniconda3/8d22aeebeaab4ce824f4fd7c3e1b98c6/lib/python3.8/site-packages/sourmash/distance_utils.py", line 148, in jaccard_to_distance
    sol1 = brentq(f1, 0.0000001, 0.9999999)
  File "/home/ntpierce/miniconda3/8d22aeebeaab4ce824f4fd7c3e1b98c6/lib/python3.8/site-packages/scipy/optimize/_zeros_py.py", line 783, in brentq
    r = _zeros._brentq(f, a, b, xtol, rtol, maxiter, args, full_output, disp)
  File "/home/ntpierce/miniconda3/8d22aeebeaab4ce824f4fd7c3e1b98c6/lib/python3.8/site-packages/sourmash/distance_utils.py", line 145, in <lambda>
    f1 = lambda pest: 2.0/(2- (1-pest)**ksize ) - 1 + z_alpha * sqrt(var_direct(pest)) - jaccard
RuntimeWarning: invalid value encountered in sqrt
```

Note, when this shows up, it shows up for both `f1` and `f2`, never just one.

I re-ran a number of analyses and recorded incidence of errors. Sent csv via email!

**Comment by mahmudhera** [comment]
I did this PR #1860, which adds codes for jaccard to point-estimate into your branch. I tested the obvious corner cases, seem to be passing. Just to note: this code does have some error, and therefore also returns an error measure with the point-estimate. More details are in the code docstrings. Let me know if you have any questions. I also added code to calculate the likelihood that nothing may be common in sketches. Look at dist_utils.py for more details.

**Comment by bluegenes** [comment]
This PR was split & refactored into 

- https://github.com/sourmash-bio/sourmash/pull/1934
- https://github.com/sourmash-bio/sourmash/pull/1952 
- https://github.com/sourmash-bio/sourmash/pull/1955 
- https://github.com/sourmash-bio/sourmash/pull/1966
- https://github.com/sourmash-bio/sourmash/pull/1967

The only remaining functionality needed is adding ANI thresholding to `tax` summarization.

**Comment by bluegenes** [comment]
punted tax ANI to #2005 and JOSS updates to #2006. I think this can safely be closed now!
